File: firebase/functions/src/utils/search.utils.test.ts
import { removeAccents, generateKeywords } from './search.utils';

describe('removeAccents', () => {
  test('removes common Portuguese accents', () => {
    expect(removeAccents('João')).toBe('Joao');
    expect(removeAccents('José')).toBe('Jose');
    expect(removeAccents('Açúcar')).toBe('Acucar');
    expect(removeAccents('Ação')).toBe('Acao');
    expect(removeAccents('Café')).toBe('Cafe');
  });

  test('removes Spanish accents', () => {
    expect(removeAccents('Señor')).toBe('Senor');
    expect(removeAccents('Niño')).toBe('Nino');
    expect(removeAccents('Español')).toBe('Espanol');
  });

  test('preserves case', () => {
    expect(removeAccents('JOÃO')).toBe('JOAO');
    expect(removeAccents('JoÃo')).toBe('JoAo');
  });

  test('preserves non-accented characters', () => {
    expect(removeAccents('abc')).toBe('abc');
    expect(removeAccents('123')).toBe('123');
    expect(removeAccents('test')).toBe('test');
  });
});

describe('generateKeywords', () => {
  test('removes accents and special characters', () => {
    expect(generateKeywords('João Da Silva&*-')).toEqual(['joao', 'da', 'silva']);
  });

  test('handles names with accents', () => {
    expect(generateKeywords('Maria José')).toEqual(['maria', 'jose']);
  });

  test('removes special characters', () => {
    expect(generateKeywords('Açúcar & Café!')).toEqual(['acucar', 'cafe']);
  });

  test('handles null input', () => {
    expect(generateKeywords(null)).toEqual([]);
  });

  test('handles undefined input', () => {
    expect(generateKeywords(undefined)).toEqual([]);
  });

  test('handles empty string', () => {
    expect(generateKeywords('')).toEqual([]);
  });

  test('preserves numbers', () => {
    expect(generateKeywords('123 ABC')).toEqual(['123', 'abc']);
  });

  test('handles multiple spaces', () => {
    expect(generateKeywords('João    Maria')).toEqual(['joao', 'maria']);
  });

  test('handles leading and trailing spaces', () => {
    expect(generateKeywords('  João Silva  ')).toEqual(['joao', 'silva']);
  });

  test('handles only special characters', () => {
    expect(generateKeywords('&*@#$%')).toEqual([]);
  });

  test('handles real world examples - Brazilian names', () => {
    expect(generateKeywords('José da Silva Júnior')).toEqual([
      'jose',
      'da',
      'silva',
      'junior',
    ]);
    expect(generateKeywords('Maria Conceição dos Santos')).toEqual([
      'maria',
      'conceicao',
      'dos',
      'santos',
    ]);
  });

  test('handles real world examples - Company names', () => {
    expect(generateKeywords('Oficina do João - ME')).toEqual([
      'oficina',
      'do',
      'joao',
      'me',
    ]);
    expect(generateKeywords('Auto Peças São Paulo Ltda.')).toEqual([
      'auto',
      'pecas',
      'sao',
      'paulo',
      'ltda',
    ]);
  });

  test('handles real world examples - Product/Service names', () => {
    expect(generateKeywords('Troca de Óleo 5W30')).toEqual([
      'troca',
      'de',
      'oleo',
      '5w30',
    ]);
  });

  test('handles uppercase input', () => {
    expect(generateKeywords('JOÃO DA SILVA')).toEqual(['joao', 'da', 'silva']);
  });

  test('handles mixed case input', () => {
    expect(generateKeywords('JoÃo Da SiLvA')).toEqual(['joao', 'da', 'silva']);
  });
});

File: test/utils/search_utils_test.dart
import 'package:praticos/utils/search_utils.dart';
import 'package:test/test.dart';

void main() {
  group('removeAccents', () {
    test('removes common Portuguese accents', () {
      expect(removeAccents('João'), equals('joao'));
      expect(removeAccents('José'), equals('jose'));
      expect(removeAccents('Açúcar'), equals('acucar'));
      expect(removeAccents('Ação'), equals('acao'));
      expect(removeAccents('Café'), equals('cafe'));
    });

    test('removes Spanish accents', () {
      expect(removeAccents('Señor'), equals('senor'));
      expect(removeAccents('Niño'), equals('nino'));
      expect(removeAccents('Español'), equals('espanol'));
    });

    test('handles mixed case', () {
      expect(removeAccents('JOÃO'), equals('joao'));
      expect(removeAccents('JoÃo'), equals('joao'));
    });

    test('preserves non-accented characters', () {
      expect(removeAccents('abc'), equals('abc'));
      expect(removeAccents('123'), equals('123'));
      expect(removeAccents('test'), equals('test'));
    });
  });

  group('generateKeywords', () {
    test('removes accents and special characters', () {
      expect(
        generateKeywords('João Da Silva&*-'),
        equals(['joao', 'da', 'silva']),
      );
    });

    test('handles names with accents', () {
      expect(
        generateKeywords('Maria José'),
        equals(['maria', 'jose']),
      );
    });

    test('removes special characters', () {
      expect(
        generateKeywords('Açúcar & Café!'),
        equals(['acucar', 'cafe']),
      );
    });

    test('handles null input', () {
      expect(generateKeywords(null), equals([]));
    });

    test('handles empty string', () {
      expect(generateKeywords(''), equals([]));
    });

    test('preserves numbers', () {
      expect(
        generateKeywords('123 ABC'),
        equals(['123', 'abc']),
      );
    });

    test('handles multiple spaces', () {
      expect(
        generateKeywords('João    Maria'),
        equals(['joao', 'maria']),
      );
    });

    test('handles leading and trailing spaces', () {
      expect(
        generateKeywords('  João Silva  '),
        equals(['joao', 'silva']),
      );
    });

    test('handles only special characters', () {
      expect(
        generateKeywords('&*@#\$%'),
        equals([]),
      );
    });

    test('handles real world examples', () {
      // Brazilian names
      expect(
        generateKeywords('José da Silva Júnior'),
        equals(['jose', 'da', 'silva', 'junior']),
      );
      expect(
        generateKeywords('Maria Conceição dos Santos'),
        equals(['maria', 'conceicao', 'dos', 'santos']),
      );

      // Company names
      expect(
        generateKeywords('Oficina do João - ME'),
        equals(['oficina', 'do', 'joao', 'me']),
      );
      expect(
        generateKeywords('Auto Peças São Paulo Ltda.'),
        equals(['auto', 'pecas', 'sao', 'paulo', 'ltda']),
      );

      // Product/Service names
      expect(
        generateKeywords('Troca de Óleo 5W30'),
        equals(['troca', 'de', 'oleo', '5w30']),
      );
    });

    test('handles uppercase input', () {
      expect(
        generateKeywords('JOÃO DA SILVA'),
        equals(['joao', 'da', 'silva']),
      );
    });

    test('handles mixed case input', () {
      expect(
        generateKeywords('JoÃo Da SiLvA'),
        equals(['joao', 'da', 'silva']),
      );
    });
  });
}
